applications/StructuralMechanicsApplication/python_scripts/structural_response.py

```diff
@@ -220,7 +220,215 @@ def GetShapeGradient(self):
+        self.ProjectParametersPrimal = Parameters( open(project_parameters["primal_settings"].GetString(),'r').read() )
+        self.primal_analysis = structural_mechanics_analysis.StructuralMechanicsAnalysis(model, self.ProjectParametersPrimal)
+
+        self.primal_model_part.AddNodalSolutionStepVariable(SHAPE_SENSITIVITY)
+
+        # Add variables to save the solution of the adjoint problem
+        self.primal_model_part.AddNodalSolutionStepVariable(StructuralMechanicsApplication.ADJOINT_DISPLACEMENT)
+        if self.ProjectParametersPrimal["solver_settings"]["rotation_dofs"].GetBool():
+            self.primal_model_part.AddNodalSolutionStepVariable(StructuralMechanicsApplication.ADJOINT_ROTATION)
+        #self.primal_model_part.AddNodalSolutionStepVariable(StructuralMechanicsApplication.POINT_LOAD_SENSITIVITY)
+        # TODO: Is it necessary to add other variables (e.g. POINT_LOAD_SENSITIVITY)?
+
+        self.primal_model_part.ProcessInfo[StructuralMechanicsApplication.IS_ADJOINT] = False
+
+    def Initialize(self):
+        self.primal_analysis.Initialize()
+
+    def InitializeSolutionStep(self):
+        print("\n> Starting primal analysis for response:", self.identifier)
+        startTime = timer.time()
+        self.primal_analysis.time = self.primal_analysis._GetSolver().AdvanceInTime(self.primal_analysis.time)
+        self.primal_analysis.InitializeSolutionStep()
+        self.primal_analysis._GetSolver().Predict()
+        self.primal_analysis._GetSolver().SolveSolutionStep()
+        self.primal_analysis.FinalizeSolutionStep()
+        self.primal_analysis.OutputSolutionStep()
+        print("> Time needed for solving the primal analysis = ",round(timer.time() - startTime,2),"s")
+
+    def CalculateValue(self):
+        startTime = timer.time()
+        value = self.response_function_utility.CalculateValue(self.primal_model_part)
+        print("> Time needed for calculating the response value = ",round(timer.time() - startTime,2),"s")
+        self.primal_model_part.ProcessInfo[StructuralMechanicsApplication.RESPONSE_VALUE] = value
+
+    def CalculateGradient(self):
+        # Replace elements and conditions by its adjoint equivalents
+        self.__PerformReplacementProcess()
+        self.response_function_utility.Initialize()
+
+        # 'Solve' the adjoint problem
+        for node in self.primal_model_part.Nodes:
+            adjoint_displacement = 0.5 * node.GetSolutionStepValue(DISPLACEMENT)
+            node.SetSolutionStepValue(StructuralMechanicsApplication.ADJOINT_DISPLACEMENT, adjoint_displacement)
+            if self.primal_analysis._GetSolver().settings["rotation_dofs"].GetBool():
+                adjoint_rotation = 0.5 * node.GetSolutionStepValue(ROTATION)
+                node.SetSolutionStepValue(StructuralMechanicsApplication.ADJOINT_ROTATION, adjoint_rotation)
+
+        # Compute Sensitivities in a post-processing step
+        self.response_function_utility.FinalizeSolutionStep()
```

Comment:
Wrong place to call FinalizeSolutionStep(). I think it is more appropriate to call UpdateSensitivities() and eventually replace this by a separate sensitivity builder.

Reply:
Since the `UpdateSensitivities()` will not be present in the base class of PR #2423 i would not like to add it here. So `FinalizeSolutionStep()` is the place where the gradients are calculated. It needs therefor be called here in the `CalculateGradient()` function.

Reply:
My argument here is the same as in my last comment. The condensed version is, `InitializeSolutionStep` ... `FinalizeSolutionStep` should either be called together in the same function body, or not at all. Only exception to this should be calling, e.g., FinalizeSolutionStep from within another FinalizeSolutionStep. If what you need does not fit this pattern (because you require a different logic), it is better to name your functions differently. One option would be to include UpdateSensitivities in this PR and split this into a different module (e.g., sensitivity builder) later on or move this directly to another module. At a minimum (and only because we are trying to resolve this PR before github breaks) a TODO comment should be inserted to replace these calls by one of the options above.

Reply:
This is part of the requested changes of the current review.

Reply:
`UpdateSensitivities()` is now exposed to python and called in the adjoint solver